firebase_admin/mlkit.py

```diff
@@ -459,6 +558,93 @@ def __init__(self, app):
+        """Sleeps for the appropriate amount of time. Or thows deadline exceeded."""
+        delay_factor = pow(
+            _MLKitService.POLL_EXPONENTIAL_BACKOFF_FACTOR, current_attempt)
+        wait_time_seconds = delay_factor * _MLKitService.POLL_BASE_WAIT_TIME_SECONDS
+
+        if stop_time is not None:
+            max_seconds_left = (stop_time - datetime.datetime.now()).total_seconds()
+            if max_seconds_left < 1: # allow a bit of time for rpc
+                raise exceptions.DeadlineExceededError('Polling max time exceeded.')
+            else:
+                wait_time_seconds = min(wait_time_seconds, max_seconds_left - 1)
+        time.sleep(wait_time_seconds)
+
+
+    def handle_operation(self, operation, polling=False, max_time_seconds=None,
+                         always_return_model=True):
+        """Handles long running operations.
+
+        Args:
+            operation: The operation to handle.
+            polling: Should we allow polling for the operation to complete.
+            max_time_seconds: The maximum seconds to try polling for operation complete.
+                (None for no limit)
+            always_return_model: If true, returns a locked Model instead of raising deadline
+                exceeded exceptions.
+
+        Returns:
+            dict: A dictionary of the returned model properties.
+
+        Raises:
+            TypeError: if the operation is not a dictionary.
+            ValueError: If the operation is malformed.
+            err: If the operation exceeds polling attempts or stop_time
+        """
+        if not isinstance(operation, dict):
+            raise TypeError('Operation must be a dictionary.')
+        op_name = operation.get('name')
+        _, model_id = _validate_and_parse_operation_name(op_name)
+
+        current_attempt = 0
+        start_time = datetime.datetime.now()
+        stop_time = (None if max_time_seconds is None else
+                     start_time + datetime.timedelta(seconds=max_time_seconds))
+        while polling and not operation.get('done'):
+            # We just got this operation. Wait before getting another
+            # so we don't exceed the GetOperation maximum request rate.
+            try:
+                self._exponential_backoff(current_attempt, stop_time)
+            except exceptions.DeadlineExceededError as err:
+                if always_return_model:
+                    return get_model(model_id).as_dict()
+                raise err
+            operation = self.get_operation(op_name)
+            current_attempt += 1
+
+        if operation.get('done'):
+            if operation.get('response'):
+                return operation.get('response')
+            elif operation.get('error'):
+                raise _utils.handle_operation_error(operation.get('error'))
+            else:
+                # A 'done' operation must have either a response or an error.
+                raise ValueError('Operation is malformed.')
```

Comment:
Push this out of the if-else structure so it becomes the default exit clause. Otherwise `done=false and always_return_model=false` case goes unhandled. This probably doesn't occur in practice, but it's good to cover all possible branches. Also may be change the error message:

```
raise ValueError('Invalid operation response: {0}`.format(operation))
```

Reply:
removing always_return_model instead and making it the opposite of polling essentially. So we either have a done operation, a deadline exceeded, or (no polling) and return a locked model.